lib/src/main/java/io/ably/lib/util/AblyErrors.java

```diff
@@ -0,0 +1,189 @@
+    FAIL_RECOVER_CONNECTION_INCOMPATIBLE_AUTH_PARAMS(80011, "unable to recover connection (incompatible auth params)"),
+    FAIL_RECOVER_CONNECTION_INVALID_CONNECTION_SERIAL(80012, "unable to recover connection (invalid or unspecified connection serial)"),
+    PROTOCOL_ERROR(80013, "protocol error"),
+    CONNECTION_TIMEOUT(80014, "connection timed out"),
+    INCOMPATIBLE_CONNECTION_PARAMS(80015, "incompatible connection parameters"),
+    OPERATION_ON_SUPERSEDED_CONNECTION(80016, "operation on superseded connection"),
+    CONNECTION_CLOSED(80017, "connection closed"),
+    INVALID_CONNECTION_ID_BAD_FORMAT(80018, "invalid connection id (invalid format)"),
+    CLIENT_AUTH_REQUEST_FAILED(80019, "client configured authentication provider request failed"),
+    CONTINUITY_LOSS_SUBSCRIBE_RATE_EXCEED(80020, "continuity loss due to maximum subscribe message rate exceeded"),
+    CREATE_CONNECTION_EXCEED(80021, "exceeded maximum permitted account-wide rate of creating new connections"),
+    RESTRICTION_NOT_SATISFIED(80030, "client restriction not satisfied"),
+
+    CHANNEL_OPERATION_FAILED(90000, "channel operation failed"),
+    CHANNEL_OPERATION_FAILED_INVALID_STATE(90001, "channel operation failed (invalid channel state)"),
+    CHANNEL_OPERATION_FAILED_EPOCH_EXPIRED(90002, "channel operation failed (epoch expired or never existed)"),
+    CHANNEL_RECOVER_ERROR_MESSAGE_EXPIRED(90003, "unable to recover channel (messages expired)"),
+    CHANNEL_RECOVER_ERROR_MESSAGE_LIMIT_EXCEED(90004, "unable to recover channel (message limit exceeded)"),
+    CHANNEL_RECOVER_ERROR_NO_EPOCH(90005, "unable to recover channel (no matching epoch)"),
+    CHANNEL_RECOVER_ERROR_UNBOUNDED_REQUEST(90006, "unable to recover channel (unbounded request)"),
+    CHANNEL_RECOVER_ERROR_NO_RESPONSE(90007, "channel operation failed (no response from server)"),
+    CHANNELS_PER_CONNECTION_EXCEED(90010, "maximum number of channels per connection/request exceeded"),
+    CREATE_CHANNEL_EXCEED(90021, "exceeded maximum permitted account-wide rate of creating new channels"),
+    CHANNEL_PRESENCE_ENTER_CLIENT_ID_ERROR(91000, "unable to enter presence channel (no clientId)"),
+    CHANNEL_PRESENCE_ENTER_STATE_ERROR(91001, "unable to enter presence channel (invalid channel state)"),
+    CHANNEL_PRESENCE_LEAVE_ERROR(91002, "unable to leave presence channel that is not entered"),
+    CHANNEL_PRESENCE_ENTER_LIMIT_EXCEED(91003, "unable to enter presence channel (maximum member limit exceeded)"),
+    CHANEL_PRESENCE_RE_ENTER_ERROR(91004, "unable to automatically re-enter presence channel"),
+    PRESENCE_STATE_BAD_SYNC(91005, "presence state is out of sync"),
+    MEMBER_LEFT_IMPLICITLY(91100, "member implicitly left presence channel (connection closed)");
+
+    public final int code;
+    public final String message;
+    private static final Map<Integer, AblyErrors> BY_CODE = new HashMap<>();
+    private static final Map<String, AblyErrors> BY_MESSAGE = new HashMap<>();
```

Comment:
I can't see any of these (valueOf methods that use`BY_CODE`, BY_MESSAGE) are used in codebase - Can you remove them?

Reply:
I have removed it https://github.com/ably/ably-java/pull/876/commits/15aff9046985575fc8642c138a86395b41c71eea